### ISSUE TITLE
Remove/Cancel Live-VOD archive from queue 

### ISSUE THREAD

**Issue by Aerglonus** [Zibbp/ganymede#51]
Is there a way to remove/cancel from queue a Live Archive/ VOD Archive in process? 

I noticed that sometimes because the "Download Thumbnail" process fails I get a duplicate external ID but different ID "Live Archive" queue process and both processes are stuck in the "Download Thumbnail" and I manually restart the stop to fix the live archive, but one queue process still there stuck at a "Failed download thumbnail" so the only workaround to remove it from the queue what I've done is manually deleting the row from the database.

**Comment by Zibbp** [comment]
Hi,
If you go to Admin > Queue you should be able to delete the queue item there. Do note, this does not stop any active queue jobs, it simply removes the queue row from the database.

I'm curious as to why the thumbnail is failing to download. I'm assuming that you have your "check live interval" set it picks up the live stream close to when the broadcaster starts? If that's the case Twitch hasn't generated the thumbnail for the stream yet, but it should not fail as Twitch provides a "blank" thumbnail if one doesn't exist.

If possible, can you post the error that gets logged to the container logs?

**Comment by Aerglonus** [comment]
> If you go to Admin > Queue you should be able to delete the queue item there. Do note, this does not stop any active queue jobs, it simply removes the queue row from the database.

So if I remove and active queue item from there, it should automatically stop the job, right? 

> I'm curious as to why the thumbnail is failing to download. I'm assuming that you have your "check live interval" set to a lower number so it picks up the live stream close to when the broadcaster starts? If that's the case Twitch hasn't generated the thumbnail for the stream yet, but it should not fail as Twitch provides a "blank" thumbnail if one doesn't exist.
> If possible, can you post the error that gets logged to the container logs?

I made the mistake of redeploying the container without saving the logs but after redeploying the container I made some changes to the `check live interval` from 5 seconds to 10 seconds, and I wasn't able to replicate the error, but today I got a small internet disconnection (this didn't happen on the previous error) for less than a minute while one queue job was running causing it to create a new queue job with the same EXT ID but different database ID, the new job triggered the `failed to download thumbnail` (keep in mind stream was already live for around an hour) stayed there until I manually restarted the `download thumbnail` step and starter downloading a new file, the old queue job resumed the `video download` except for the `chat download` which exited with an error and couldn't restart it. 

Is it possible that if the new queue job gets deleted, the old queue video file will turn out incomplete? Even though they are shown as two different files? 

CONTAINER ERROR CAUSED BY THE INTERNET OUTAGE:
> {"level":"error","error":"error downloading file: Get \"https://static-cdn.jtvnw.net/previews-ttv/live_user_sushii-1920x1080.jpg\": net/http: TLS handshake timeout","time":"2022-11-03T08:05:13Z","message":"error downloading thumbnail"}

DISCORD NOTIFICATION : 
> Error: Queue ID 2e292f27-6ecb-44a4-8868-6fe3570aaa7c failed at task vod_download_thumbnail.

FIRST ITEM ON THE QUEUE IS THE JOB CREATED AFTER INTERNET OUTAGE `2e292f27-6ecb-44a4-8868-6fe3570aaa7c`

![image](https://user-images.githubusercontent.com/1855681/199675192-e05dfa58-96aa-4f44-8493-ff1a8af4d6b0.png)

**Comment by Zibbp** [comment]
> I made some changes to the check live interval from 5 seconds to 10 seconds

From my experience such a low check interval such as 5 seconds causes issues how the Twitch API updates if a stream is online. 10 or 15 seconds is preferred.

> I got a small internet disconnection (this didn't happen on the previous error) for less than a minute while one queue job was running causing it to create a new queue job with the same EXT ID but different database ID

This is intentional. Losing internet should cause the video and chat download to exit. A new queue job gets created for the live stream as the previous one has stopped archiving.

> the new job triggered the failed to download thumbnail (keep in mind stream was already live for around an hour) stayed there until I manually restarted the download thumbnail step and starter downloading a new file

> {"level":"error","error":"error downloading file: Get "[https://static-cdn.jtvnw.net/previews-ttv/live_user_sushii-1920x1080.jpg\](https://static-cdn.jtvnw.net/previews-ttv/live_user_sushii-1920x1080.jpg%5C)": net/http: TLS handshake timeout","time":"2022-11-03T08:05:13Z","message":"error downloading thumbnail"}

I believe the thumbnail failed to download due to your internet slowly coming back up or another brief interruption *after* the main interruption.

> the old queue job resumed the video download except for the chat download which exited with an error and couldn't restart it.

[Streamlink](https://streamlink.github.io/) (the video downloading application) takes 20-45 seconds to realize that you lost internet connection. During this 20-45 second period it will show errors in the logs *but* will not exit the application. I'm guessing the reason why it continued to download after your internet came back up is because it did not fully exit yet. It was waiting the 20-45 seconds and once your internet came up it continued to download.

I'm not sure why the chat download for the first queue item failed. The application, [chat-downloader](https://github.com/xenova/chat-downloader), does not exit the application if internet connection is lost, it will just hang not doing anything. In Ganymede the way chat-downloader stops is when the video download is complete or errors it will then kill the chat download.

Was there any logs relating to the chat download failing? I'm guessing not as the application does not throw any.

> Is it possible that if the new queue job gets deleted, the old queue video file will turn out incomplete? Even though they are shown as two different files?

The intended way of handling interruptions is the first queue job will exit and process. Then once internet is restored a new queue job gets created. This results with two VOD entries in Ganymede, the first being `streamer start time` to `internet outage` the second being `internet outage` to `streamer end time`.

There is no logic in Ganymede to handle two different video files from Streamlink in a single queue so I will assume that it will be incomplete. The first queue item with have the video up until your internet crashed and the second queue item will have the video starting from when your internet was restored.

Hopefully that all makes sense.

**Question** It seems that the Thumbnail download job causes issues quite often. Would it be beneficial to allow the thumbnail download job to fail but not block the rest of the queue? Meaning the next step can continue if the thumbnail job fails.

**Comment by Zibbp** [comment]
On a somewhat related note, it appears Twitch has released a beta for their websocket EventSub API https://dev.twitch.tv/docs/eventsub. Previously the only way to access the EventSub API was using webhooks which required a publicly accessible callback from a server (I didn't to require everyone running Ganymede to have it publicly accessible). Now with websockets that isn't required. The EventSub API has a "stream online" event which is probably much better than hitting the Twitch API every X seconds to check if a stream is live. I'll have to investigate this some more.

**Comment by Aerglonus** [comment]
> Was there any logs relating to the chat download failing? I'm guessing not as the application does not throw any.

(`Note`: for this exact same job the VoD file is fully complete, no cuts or missing frame at the timestamp the outage happened (total 5 hours of stream, outage happened at 2 hours in) except for the chat video that obviously ends at the time the outage happened also noticed that even though the `chat download` exited the `chat convert` triggered until the stream ended like it normally would)

For the queue job that was running during the internet outage and the `video download` continued but `chat download` failed, the chat log file shows this :


```
Chat downloader started. It it unlikely that you will see further output in this log.Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/twitch.py", line 1553, in _get_chat_messages_by_stream_id
    raise ConnectionError('Lost connection, reconnecting.')
ConnectionError: Lost connection, reconnecting.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/twitch.py", line 1528, in create_connection
    irc = TwitchChatIRC()
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/twitch.py", line 57, in __init__
    self.socket.connect(('irc.chat.twitch.tv', 6667))
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/chat_downloader", line 8, in <module>
    sys.exit(main())
  File "/usr/lib/python3.10/site-packages/chat_downloader/cli.py", line 194, in main
    run(**args.__dict__)
  File "/usr/lib/python3.10/site-packages/chat_downloader/chat_downloader.py", line 360, in run
    for message in chat:
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/common.py", line 283, in __next__
    item = next(self.chat)
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/twitch.py", line 1649, in _get_chat_messages_by_stream_id
    twitch_chat_irc = create_connection()
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/twitch.py", line 1533, in create_connection
    self.retry(attempt_number, error=e, **params)
  File "/usr/lib/python3.10/site-packages/chat_downloader/sites/common.py", line 648, in retry
    timed_input(time_to_sleep)
  File "/usr/lib/python3.10/site-packages/chat_downloader/utils/timed_utils.py", line 86, in timed_input
    return _timed_input(timeout, prompt, newline)
  File "/usr/lib/python3.10/site-packages/chat_downloader/utils/timed_utils.py", line 67, in posix_timed_input
    sel.register(sys.stdin, selectors.EVENT_READ)
  File "/usr/lib/python3.10/selectors.py", line 360, in register
    self._selector.register(key.fd, poller_events)
PermissionError: [Errno 1] Operation not permitted
```
> Question It seems that the Thumbnail download job causes issues quite often. Would it be beneficial to allow the thumbnail download job to fail but not block the rest of the queue? Meaning the next step can continue if the thumbnail job fails.

Totally, so far the discord webhook it's helpful but at the same time so easy to miss the notification.

**Comment by Zibbp** [comment]
Released [v1.0.7](https://github.com/Zibbp/ganymede/releases/tag/v1.0.7) which should allow the live thumbnail download job to fail without blocking.

I'll look more into the video and chat download issue when losing internet connection but I don't think there's much I can do. The current implementation of allowing the video download to fail then kill the chat download I believe is the best. There will always be an unfortunate small window of time if your internet comes back in time that it may not kill the first queue item while creating a new one. I'll take a look and see if there's any improvements that can be made.

**Comment by Zibbp** [comment]
Closing due to inactivity.
